transaction/std/src/memo_builder/rth_memo_builder.rs

```diff
@@ -0,0 +1,212 @@
+/// address will be recorded in the 0x0200 Destination Memo.
+#[derive(Clone, Debug)]
+pub struct RTHMemoBuilder {
+    // The credential used to form 0x0100 and 0x0101 memos, if present.
+    sender_cred: Option<SenderMemoCredential>,
+    // The payment request id, if any
+    payment_request_id: Option<u64>,
+    // Whether destination memos are enabled.
+    destination_memo_enabled: bool,
+    // Tracks if we already wrote a destination memo, for error reporting
+    wrote_destination_memo: bool,
+    // Tracks the last recipient
+    last_recipient: AddressHash,
+    // Tracks the total outlay so far
+    total_outlay: u64,
+    // Tracks the number of recipients so far
+    num_recipients: u8,
+    // Tracks the fee
+    fee: u64,
+}
+
+impl Default for RTHMemoBuilder {
+    fn default() -> Self {
+        Self {
+            sender_cred: Default::default(),
+            payment_request_id: None,
+            destination_memo_enabled: false,
+            wrote_destination_memo: false,
+            last_recipient: Default::default(),
+            total_outlay: 0,
+            num_recipients: 0,
+            fee: MINIMUM_FEE,
+        }
+    }
+}
+
+impl RTHMemoBuilder {
+    /// Set the sender credential. If no sender credential is provided,
+    /// then authenticated sender memos cannot be produced.
+    ///
+    /// This credential usually be produced from your AccountKey object.
+    ///
+    /// If you want to make it appear to the recipient as if this came from
+    /// another address or a subaddress of yours,
+    /// you can do that as long as you have the spend private key
+    /// for that address.
+    ///
+    /// For example, if you are an exchange, and you are sending
+    /// MOB to a user, you might set this to match the subaddress that they
+    /// normally deposit to. Then a chat client will be able to associate both
+    /// their deposits and withdrawals into a single chat interaction.
+    pub fn set_sender_credential(&mut self, cred: SenderMemoCredential) {
+        self.sender_cred = Some(cred);
+    }
+
+    /// Clear the sender credential.
+    pub fn clear_sender_credential(&mut self) {
+        self.sender_cred = None;
+    }
+
+    /// Set the payment request id.
+    pub fn set_payment_request_id(&mut self, id: u64) {
+        self.payment_request_id = Some(id);
+    }
+
+    /// Clear the payment request id.
+    pub fn clear_payment_request_id(&mut self) {
+        self.payment_request_id = None;
+    }
+
+    /// Enable destination memos
+    pub fn enable_destination_memo(&mut self) {
+        self.destination_memo_enabled = true;
+    }
+
+    /// Disable destination memos
+    pub fn disable_destination_memo(&mut self) {
+        self.destination_memo_enabled = false;
+    }
+}
+
+impl MemoBuilder for RTHMemoBuilder {
+    /// Set the fee
+    fn set_fee(&mut self, fee: u64) -> Result<(), NewMemoError> {
+        if self.wrote_destination_memo {
```

Comment:
Should this include a check for the fee fitting inside 7 bytes?

Reply:
there's sort of like a design issue here where, i wanted the logic for that check to live in the `DestinationMemo` type, but that means the error doesn't get surfaced until we try to build the destination memo.

one way to change it would be to make `rth_memo_builder` hold an `DestinationMemo` or `Option<DestinationMemo>` instead of a bool ad a bunch of ints. it's just that that will make for slightly more code when we have like `self.num_recipients.checked_add`

Reply:
I think its fine to leave it as-is.